packages/fast-usdc-contract/src/utils/settlement-matching.ts

```diff
@@ -0,0 +1,128 @@
+/**
+ * @file Settlement-matching utilities for USDC transactions.
+ *
+ * This module maintains a `MapStore<NobleAddress, PendingTx[]>` where pending transactions
+ * for each Noble address are kept in **descending order** by `tx.amount`. This sorting
+ * invariant enables efficient resolution of new mints through two approaches:
+ *   1. Exact-amount direct lookup (O(n) time complexity)
+ *   2. Greedy combination algorithm that processes transactions from largest to smallest
+ *
+ * The greedy strategy is deterministic and auditable: it always selects the largest
+ * transaction that doesn't exceed the remaining target until the full amount is matched.
+ * If no valid combination exists, an empty array is returned.
+ *
+ * Implementation includes memoization and depth-limiting to ensure performance even with
+ * larger transaction sets, though typical match depth is 2-3 transactions.
+ */
+
+import type { MapStore } from '@agoric/swingset-liveslots';
+import type { NobleAddress, PendingTx } from '@agoric/fast-usdc/src/types.ts';
+import { appendToSortedStoredArray } from './store.ts';
+
+/**
+ * Max number of txs to settle against a single mint.
+ * Much higher than observed value of 2-3.
+ */
+const MAX_MATCH_DEPTH = 25;
+
+/** bigint‑safe descending comparator (largest first) */
+const comparePendingTxDesc = (a: PendingTx, b: PendingTx): number => {
+  if (a.tx.amount < b.tx.amount) return 1;
+  if (a.tx.amount > b.tx.amount) return -1;
+  return 0;
+};
+
+harden(comparePendingTxDesc);
+
+/**
+ * Depth‑first, memoised, largest‑first chooser.
+ * Returns the *first* exact combination or [] if none.
+ */
+const greedyMatch = (
+  pending: readonly PendingTx[],
+  target: bigint,
+): PendingTx[] => {
+  if (target === 0n) return [];
+  if (pending.length === 0) return [];
+
+  const memo = new Map<string, boolean>();
+  const path: PendingTx[] = [];
+
+  const dfs = (index: number, sum: bigint): boolean => {
+    if (sum === target) return true;
+    if (sum > target) return false; // unreachable with current continue‑on‑overshoot logic
+    if (path.length >= MAX_MATCH_DEPTH) return false;
```

Comment:
very unexpected to reach this depth. throw instead?

Reply:
Agree, much better idea.